Fixed null date issue and stream error handling

jd-alexander opened the pull request:
Stream errors weren't being handled so a crash was being caused when an error occurs within the stream so now that is being handled and sent to Crashlytics via the Timber log. The `SmsEvent` now has a default date that can be used by events. Events without the date were causing a crash when the formatter got a null date.

Closes #2401 

#### What has been done to verify that this works as intended?
Tested the scenario described in the issue on a physical tablet and an emulator. Crash won't be caused in the event anything weird ever happens since the stream handles errors.

#### Why is this the best possible solution? Were any other approaches considered?
It's the best solution since the stream being broken by the error was causing the crash so implemented exception handling causes this to not occur.

#### Are there any risks to merging this code? If so, what are they?
No.

#### Do we need any specific form for testing your changes? If so, please attach one.
Use a SMS Test sample form from the server.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
N/A

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

jd-alexander committed:
Fixed null date issue and stream error handling

Stream errors weren't being handled so a crash was being caused when an error occurs within the stream so now that is being handled and sent to Crashlytics via the Timber log. The SmsEvent now has a default date that can be used by events.

jd-alexander committed:
RxRelay doesn't support onError

Removed onError implementation and caught the exception the normal way due to RxRelay not supporting it so that it can have a stream that can't be terminated.

lognaturel commented:
👍 If there were a null date somehow with this code, the subtext would just be blank, is that right?

jd-alexander commented:
No, because there are no cases where it can be null really. The error occurred due to a missing date which won't ever happen again unless `new Date()` returns null. Can that be null based on your experience?

jd-alexander commented:
Just double checked and saw that if an error occurs it's returning an empty string for the subtext. What I will do is rework that to catch this exception as well.

lognaturel commented:
Are there other error states you've uncovered, then? In general, it's best to wrap as few lines as possible in a try/catch and only catch an exception that's known to occur. If you keep the try/catch, do you really need to wrap all those lines? Can you catch an NPE rather than `Exception`?

Could the subtext method be made a little more robust by checking for a null date there and replacing it by some placeholder? Would "" work or not?

jd-alexander committed:
getDisplaySubText now catches a general Exception

This IllegalArgumentException was setup for exceptions from language translations but since the date issue occurred it's now catching a general Exception so others can cause the return of the empty status.

shobhitagarwal1612 commented:
I believe there is an overriden implementation of `subscribe` method in which the second parameter is to catch all exceptions instead of throwing that might occur during stream flow. So that would remove the usage of try catch block

jd-alexander commented:
> Are there other error states you've uncovered, then? In general, it's best to wrap as few lines as possible in a try/catch and only catch an exception that's known to occur. If you keep the try/catch, do you really need to wrap all those lines? Can you catch an NPE rather than Exception?

No, I don't! Was just removing that.

jd-alexander commented:
> Could the subtext method be made a little more robust by checking for a null date there and replacing it by some placeholder? Would "" work or not?

Yes, I was catching an `IllegalArgumentException` due to the point made by @grzesiek2010 about translations throwing off switch statements. 

``` 
default:
             return new SimpleDateFormat(context.getString(R.string.sms_fatal_error), Locale.getDefault()).format(date);
            }
        } catch (IllegalArgumentException e) {
            Timber.e(e);
        }

        return "";
    } 
```

I know there's a general consensus about being careful about catching general exceptions so stuff like `OutOfMemoryException` and possibly masking other issues, so yes, I am going to just check if the date is null and return the empty string and leave this exception handling for the possible translation issues.

lognaturel commented:
Sounds good. I was wondering whether there’s a value for the Date that would make it possible to still get the subtext, just without the date. So you’d have “SMS failed on” with no date rather than no subtext at all.

jd-alexander commented:
@shobhitagarwal1612 `RxRelay` doesn't support `onError` because that breaks the stream and puts it in an unusable state and that would be problematic for uses of this event bus so you have to handle errors differently. One mechanism for doing that is passing errors within the event being pushed through the relay or simply just handling it yourself.

jd-alexander commented:
> Sounds good. I was wondering whether there’s a value for the Date that would make it possible to still get the subtext, just without the date. So you’d have “SMS failed on” with no date rather than no subtext at all.

A null date being introduced is highly unlikely especially with the cases having dates eg. submission always get's a new date and the event has a date as well for events that are created before a submission event happens.  So would returning a `SMS failed.` message be wrong? I would simply just log the null date as an error and track it down because it's really not supposed to happen.

lognaturel commented:
> I would simply just log the null date as an error and track it down because it's really not supposed to happen.

That's fair.

jd-alexander committed:
Proper handling of null date in getDisplaySubText

jd-alexander commented:
Removing the sim card from the device before testing should reproduce the issue.

lognaturel commented:
This can be hard to reproduce consistently and I've verified that once I can reproduce it, the fix does address the problem. Since it's so targeted, I don't see any risk so I will merge without an additional QA pass.

kkrawczyk123 commented:
I am not able to reproduce this issue on the newest master which is good but I wasn't able to reproduce it on master before those changes were merged and on beta from google play too which is not that good.

lognaturel commented:
@kkrawczyk123 was this with brand new forms or existing ones? I suspect that there's an issue with using stale status info for an existing form though I haven't tried to find that yet.

kkrawczyk123 commented:
I tried again with a completely new form which I created by myself. I still couldn't reproduce an issue I don't know why as most of the errors and crashes like me.

lognaturel commented:
😄 Very interesting. I don’t think we need to worry too much about it. Thanks!
